dev/com.ibm.ws.jpa.container.v32.cdi/src/com/ibm/ws/jpa/container/v32/cdi/internal/JPACDIExtension.java

```diff
@@ -0,0 +1,174 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *******************************************************************************/
+package com.ibm.ws.jpa.container.v32.cdi.internal;
+
+import java.lang.annotation.Annotation;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.ibm.websphere.csi.J2EEName;
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.jpa.JPAAccessor;
+import com.ibm.ws.jpa.JPAComponent;
+import com.ibm.ws.runtime.metadata.ComponentMetaData;
+import com.ibm.ws.threadContext.ComponentMetaDataAccessorImpl;
+
+import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.inject.spi.AfterBeanDiscovery;
+import jakarta.enterprise.inject.spi.BeanManager;
+import jakarta.enterprise.inject.spi.Extension;
+import jakarta.enterprise.util.AnnotationLiteral;
+import jakarta.persistence.Cache;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.EntityManagerFactory;
+import jakarta.persistence.PersistenceUnitUtil;
+import jakarta.persistence.SchemaManager;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.metamodel.Metamodel;
+import jakarta.persistence.spi.PersistenceUnitInfo;
+
+public class JPACDIExtension implements Extension {
+
+    private static final TraceComponent tc = Tr.register(JPACDIExtension.class);
+
+    private final JPAComponent jpaComponent = JPAAccessor.getJPAComponent();
+
+    public void afterBeanDiscovery(@Observes AfterBeanDiscovery abd, BeanManager bm) {
+
+        //JPA PersistenceUnit are scoped to modules, but in getCurrentAppJ2EEName I strip
+        //out the app name. This is because of a limitation in our CDI implementation that
+        //means we cannot scope an Extension to just a single module.
+        //
+        //This is a tech debt that really should be fixed.
+        J2EEName j2EEName = getCurrentAppJ2EEName();
+        List<PersistenceUnitInfo> persistenceUnits = jpaComponent.getPersistenceUnits(j2EEName);
+
+        for (PersistenceUnitInfo pui : persistenceUnits) {
+            try {
+                createBeanForPersistenceUnit(abd, pui, j2EEName);
+            } catch (ClassNotFoundException e) {
+                Tr.warning(tc, "FAILED_TO_CREATE_PU_BEAN_CWWJP9994W", pui.getPersistenceUnitName(), e);
+            }
+        }
+    }
+
+    private void createBeanForPersistenceUnit(AfterBeanDiscovery abd, final PersistenceUnitInfo pui, final J2EEName j2eeName) throws ClassNotFoundException {
+
+        Set<Annotation> qualifiers = getQualifiers(pui);
+        //TODO ask JPA team if anything needs to happen when any of these go out of scope.
```

Comment:
Is there an issue open for this? (We'll want to take some action, either removing the TODO or adding some code, and we don't want to forget so we should have an issue)